<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Computing evoregions analysis • Herodotools</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Computing evoregions analysis">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">Herodotools</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.1.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/bsm_analysis.html">Using Biogeographical Stochastic Mapping on Herodotools</a></li>
    <li><a class="dropdown-item" href="../articles/building-evoregions.html">Computing evoregions analysis</a></li>
    <li><a class="dropdown-item" href="../articles/community-phylogenetic-metrics.html">community-phylogenetic-metrics</a></li>
    <li><a class="dropdown-item" href="../articles/Intro_calc_insitu_dispersal.html">Calculating in situ diversification and historical dispersal</a></li>
  </ul>
</li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Computing evoregions analysis</h1>
            
      

      <div class="d-none name"><code>building-evoregions.Rmd</code></div>
    </div>

    
    
<p>This article outlines the basic workflow for constructing
<em>evoregions</em>, starting from defining the maximum number of
clusters to the visualization of the final results generated by the
<code>evoregion()</code> function. The workflow is organized into two
main sections:</p>
<ol style="list-style-type: decimal">
<li><p><a href="#setclust">Setting the maximum number of clusters</a> to
be considered in the evoregion analysis.</p></li>
<li><p><a href="#evoregion">Computing evoregions</a> using the number of
clusters that yields the most stable and interpretable solution in the
analysis.</p></li>
</ol>
<div class="section level2">
<h2 id="setclust">Setting the maximum number of clusters<a class="anchor" aria-label="anchor" href="#setclust"></a>
</h2>
<p>In the evoregion workflow, defining the maximum number of clusters to
be tested is a key step. This is particularly important because the
<code>find.clusters()</code> function from the {adegenet} package is
sensitive to this parameter: different values of the maximum number of
clusters may lead to different group assignments.</p>
<p>The purpose of the <code><a href="../reference/find_max_nclust.html">find_max_nclust()</a></code> function in
{Herodotools} is to identify the maximum number of clusters that
produces the most stable grouping solution. To do this, the function
runs <code>find.clusters()</code> multiple times across a range of
maximum cluster values and evaluates the consistency of the resulting
solutions.</p>
<p>Stability is assessed by computing the correlation between clustering
results across runs, combined with different tolerance values for the
<code>confidence.level</code> argument. The optimal maximum number of
clusters is the one that maximizes this correlation, serving as a proxy
for the stability of the evoregion solution.</p>
<p>In this tutorial, we use example data provided in the {Herodotools}
package. The <code>akodon_site</code> dataset represents the occurrence
of species across assemblages mapped in 1° × 1° grid cells. The
<code>akodon_newick</code> object contains the phylogenetic tree (in
Newick format) describing the evolutionary relationships among species
of the tribe Akodontini.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://gabrielnakamura.github.io/Herodotools/">Herodotools</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"akodon_sites"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"akodon_newick"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># minor data processing </span></span>
<span></span>
<span><span class="va">site_xy</span> <span class="op">&lt;-</span> <span class="va">akodon_sites</span>  <span class="op">|&gt;</span>  </span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">LONG</span>, <span class="va">LAT</span><span class="op">)</span> </span>
<span></span>
<span><span class="va">akodon_pa</span> <span class="op">&lt;-</span> <span class="va">akodon_sites</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="op">-</span><span class="va">LONG</span>, <span class="op">-</span><span class="va">LAT</span><span class="op">)</span></span>
<span></span>
<span><span class="va">spp_in_tree</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">akodon_pa</span><span class="op">)</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="va">akodon_newick</span><span class="op">$</span><span class="va">tip.label</span></span>
<span><span class="va">akodon_pa_tree</span> <span class="op">&lt;-</span> <span class="va">akodon_pa</span><span class="op">[</span>, <span class="va">spp_in_tree</span><span class="op">]</span></span></code></pre></div>
<p>First, we need to compute the Principal Component of Phylogenetic
Structure (PCPS) matrix, which will be used as input for the
<code><a href="../reference/find_max_nclust.html">find_max_nclust()</a></code> function. This can be done using the
<code>pcps()</code> function from the {PCPS} package.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pcps_bray</span> <span class="op">&lt;-</span></span>
<span>  <span class="fu">PCPS</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/PCPS/man/pcps.html" class="external-link">pcps</a></span><span class="op">(</span><span class="va">akodon_pa_tree</span>, phylodist <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/cophenetic.html" class="external-link">cophenetic</a></span><span class="op">(</span><span class="va">akodon_newick</span><span class="op">)</span>, method <span class="op">=</span> <span class="st">"bray"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">values_bray</span> <span class="op">&lt;-</span> <span class="va">pcps_bray</span><span class="op">$</span><span class="va">values</span> <span class="co"># PCPS eigenvalues, relative eigenvalues and cumulative relative eigenvalues</span></span>
<span></span>
<span><span class="co"># Define a threshold value for eigenvectors (eigenvectors containing more than 5% of variation)</span></span>
<span><span class="va">thresh_bray</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">max</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="va">values_bray</span><span class="op">[</span>, <span class="fl">2</span><span class="op">]</span> <span class="op">&gt;=</span> <span class="fl">0.05</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">vec_bray</span> <span class="op">&lt;-</span> <span class="va">pcps_bray</span><span class="op">$</span><span class="va">vectors</span> <span class="co"># eigenvectors </span></span></code></pre></div>
<p>Now we can estimate the optimal maximum number of groups using the
<code><a href="../reference/find_max_nclust.html">find_max_nclust()</a></code> function implemented in {Herodotools}.
This function supports parallel computation via the {future} package. We
recomend to set the parallel backend using <code><a href="https://future.futureverse.org/reference/plan.html" class="external-link">future::plan()</a></code>
first.</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" tabindex="-1"></a><span class="co"># setting function to work in parallel according to user settings</span></span>
<span id="cb3-2"><a href="#cb3-2" tabindex="-1"></a><span class="fu">library</span>(future)</span>
<span id="cb3-3"><a href="#cb3-3" tabindex="-1"></a><span class="fu">library</span>(progressr)</span>
<span id="cb3-4"><a href="#cb3-4" tabindex="-1"></a></span>
<span id="cb3-5"><a href="#cb3-5" tabindex="-1"></a><span class="co"># Detect safe max cores</span></span>
<span id="cb3-6"><a href="#cb3-6" tabindex="-1"></a>ncores <span class="ot">&lt;-</span> future<span class="sc">::</span><span class="fu">availableCores</span>()  <span class="co"># dynamic</span></span>
<span id="cb3-7"><a href="#cb3-7" tabindex="-1"></a>safety_margin <span class="ot">&lt;-</span> <span class="dv">1</span></span>
<span id="cb3-8"><a href="#cb3-8" tabindex="-1"></a>workers <span class="ot">&lt;-</span> <span class="fu">max</span>(<span class="dv">1</span>, ncores <span class="sc">-</span> safety_margin)</span>
<span id="cb3-9"><a href="#cb3-9" tabindex="-1"></a></span>
<span id="cb3-10"><a href="#cb3-10" tabindex="-1"></a><span class="fu">plan</span>(multisession, <span class="at">workers =</span> workers) this </span>
<span id="cb3-11"><a href="#cb3-11" tabindex="-1"></a>future<span class="sc">::</span><span class="fu">plan</span>(sequential)</span>
<span id="cb3-12"><a href="#cb3-12" tabindex="-1"></a><span class="fu">handlers</span>(<span class="at">global =</span> <span class="cn">TRUE</span>)</span>
<span id="cb3-13"><a href="#cb3-13" tabindex="-1"></a><span class="fu">handlers</span>(<span class="st">"txtprogressbar"</span>)   <span class="co"># terminal progress bar + timing info</span></span></code></pre></div>
<p>Then, run the <code>find_max_clust</code> function</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="va">matrix_optimal_maxclust</span> <span class="op">&lt;-</span> </span>
<span>  <span class="fu"><a href="../reference/find_max_nclust.html">find_max_nclust</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">vec_bray</span>,</span>
<span>                  threshold <span class="op">=</span> <span class="va">thresh_bray</span>, </span>
<span>                  max.nclust <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">3</span>, <span class="fl">4</span>, <span class="fl">5</span>, <span class="fl">7</span>, <span class="fl">9</span>, <span class="fl">10</span><span class="op">)</span>,</span>
<span>                  nperm <span class="op">=</span> <span class="fl">300</span>, </span>
<span>                  method <span class="op">=</span> <span class="st">"kmeans"</span>, </span>
<span>                  stat <span class="op">=</span> <span class="st">"BIC"</span>, </span>
<span>                  criterion <span class="op">=</span> <span class="st">"diffNgroup"</span>,</span>
<span>                  subset <span class="op">=</span> <span class="fl">100</span>,</span>
<span>                  confidence.level <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.7</span>, <span class="fl">0.8</span>, <span class="fl">0.9</span>, <span class="fl">0.95</span>, <span class="fl">0.99</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning: UNRELIABLE VALUE: One of the 'future.apply' iterations</span></span>
<span><span class="co">#&gt; ('future_lapply-1') unexpectedly generated random numbers without declaring so.</span></span>
<span><span class="co">#&gt; There is a risk that those random numbers are not statistically sound and the</span></span>
<span><span class="co">#&gt; overall results might be invalid. To fix this, specify 'future.seed=TRUE'. This</span></span>
<span><span class="co">#&gt; ensures that proper, parallel-safe random numbers are produced via a parallel</span></span>
<span><span class="co">#&gt; RNG method. To disable this check, use 'future.seed = NULL', or set option</span></span>
<span><span class="co">#&gt; 'future.rng.onMisuse' to "ignore".</span></span>
<span><span class="co">#&gt; Warning: UNRELIABLE VALUE: One of the 'future.apply' iterations</span></span>
<span><span class="co">#&gt; ('future_lapply-1') unexpectedly generated random numbers without declaring so.</span></span>
<span><span class="co">#&gt; There is a risk that those random numbers are not statistically sound and the</span></span>
<span><span class="co">#&gt; overall results might be invalid. To fix this, specify 'future.seed=TRUE'. This</span></span>
<span><span class="co">#&gt; ensures that proper, parallel-safe random numbers are produced via a parallel</span></span>
<span><span class="co">#&gt; RNG method. To disable this check, use 'future.seed = NULL', or set option</span></span>
<span><span class="co">#&gt; 'future.rng.onMisuse' to "ignore".</span></span>
<span><span class="co">#&gt; Warning: UNRELIABLE VALUE: One of the 'future.apply' iterations</span></span>
<span><span class="co">#&gt; ('future_lapply-1') unexpectedly generated random numbers without declaring so.</span></span>
<span><span class="co">#&gt; There is a risk that those random numbers are not statistically sound and the</span></span>
<span><span class="co">#&gt; overall results might be invalid. To fix this, specify 'future.seed=TRUE'. This</span></span>
<span><span class="co">#&gt; ensures that proper, parallel-safe random numbers are produced via a parallel</span></span>
<span><span class="co">#&gt; RNG method. To disable this check, use 'future.seed = NULL', or set option</span></span>
<span><span class="co">#&gt; 'future.rng.onMisuse' to "ignore".</span></span>
<span><span class="co">#&gt; Warning: Quick-TRANSfer stage steps exceeded maximum (= 36600)</span></span>
<span><span class="co">#&gt; Warning: UNRELIABLE VALUE: One of the 'future.apply' iterations</span></span>
<span><span class="co">#&gt; ('future_lapply-1') unexpectedly generated random numbers without declaring so.</span></span>
<span><span class="co">#&gt; There is a risk that those random numbers are not statistically sound and the</span></span>
<span><span class="co">#&gt; overall results might be invalid. To fix this, specify 'future.seed=TRUE'. This</span></span>
<span><span class="co">#&gt; ensures that proper, parallel-safe random numbers are produced via a parallel</span></span>
<span><span class="co">#&gt; RNG method. To disable this check, use 'future.seed = NULL', or set option</span></span>
<span><span class="co">#&gt; 'future.rng.onMisuse' to "ignore".</span></span>
<span><span class="co">#&gt; Warning: Quick-TRANSfer stage steps exceeded maximum (= 36600)</span></span>
<span><span class="co">#&gt; Warning: Quick-TRANSfer stage steps exceeded maximum (= 36600)</span></span>
<span><span class="co">#&gt; Warning: Quick-TRANSfer stage steps exceeded maximum (= 36600)</span></span>
<span><span class="co">#&gt; Warning: Quick-TRANSfer stage steps exceeded maximum (= 36600)</span></span>
<span><span class="co">#&gt; Warning: Quick-TRANSfer stage steps exceeded maximum (= 36600)</span></span>
<span><span class="co">#&gt; Warning: Quick-TRANSfer stage steps exceeded maximum (= 36600)</span></span>
<span><span class="co">#&gt; Warning: Quick-TRANSfer stage steps exceeded maximum (= 36600)</span></span>
<span><span class="co">#&gt; Warning: Quick-TRANSfer stage steps exceeded maximum (= 36600)</span></span>
<span><span class="co">#&gt; Warning: Quick-TRANSfer stage steps exceeded maximum (= 36600)</span></span>
<span><span class="co">#&gt; Warning: Quick-TRANSfer stage steps exceeded maximum (= 36600)</span></span>
<span><span class="co">#&gt; Warning: Quick-TRANSfer stage steps exceeded maximum (= 36600)</span></span>
<span><span class="co">#&gt; Warning: Quick-TRANSfer stage steps exceeded maximum (= 36600)</span></span>
<span><span class="co">#&gt; Warning: did not converge in 100000 iterations</span></span>
<span><span class="co">#&gt; Warning: Quick-TRANSfer stage steps exceeded maximum (= 36600)</span></span>
<span><span class="co">#&gt; Warning: UNRELIABLE VALUE: One of the 'future.apply' iterations</span></span>
<span><span class="co">#&gt; ('future_lapply-1') unexpectedly generated random numbers without declaring so.</span></span>
<span><span class="co">#&gt; There is a risk that those random numbers are not statistically sound and the</span></span>
<span><span class="co">#&gt; overall results might be invalid. To fix this, specify 'future.seed=TRUE'. This</span></span>
<span><span class="co">#&gt; ensures that proper, parallel-safe random numbers are produced via a parallel</span></span>
<span><span class="co">#&gt; RNG method. To disable this check, use 'future.seed = NULL', or set option</span></span>
<span><span class="co">#&gt; 'future.rng.onMisuse' to "ignore".</span></span>
<span><span class="co">#&gt; Warning: Quick-TRANSfer stage steps exceeded maximum (= 36600)</span></span>
<span><span class="co">#&gt; Warning: Quick-TRANSfer stage steps exceeded maximum (= 36600)</span></span>
<span><span class="co">#&gt; Warning: Quick-TRANSfer stage steps exceeded maximum (= 36600)</span></span>
<span><span class="co">#&gt; Warning: Quick-TRANSfer stage steps exceeded maximum (= 36600)</span></span>
<span><span class="co">#&gt; Warning: did not converge in 100000 iterations</span></span>
<span><span class="co">#&gt; Warning: Quick-TRANSfer stage steps exceeded maximum (= 36600)</span></span>
<span><span class="co">#&gt; Warning: did not converge in 100000 iterations</span></span>
<span><span class="co">#&gt; Warning: Quick-TRANSfer stage steps exceeded maximum (= 36600)</span></span>
<span><span class="co">#&gt; Warning: Quick-TRANSfer stage steps exceeded maximum (= 36600)</span></span>
<span><span class="co">#&gt; Warning: did not converge in 100000 iterations</span></span>
<span><span class="co">#&gt; Warning: Quick-TRANSfer stage steps exceeded maximum (= 36600)</span></span>
<span><span class="co">#&gt; Warning: did not converge in 100000 iterations</span></span>
<span><span class="co">#&gt; Warning: Quick-TRANSfer stage steps exceeded maximum (= 36600)</span></span>
<span><span class="co">#&gt; Warning: Quick-TRANSfer stage steps exceeded maximum (= 36600)</span></span>
<span><span class="co">#&gt; Warning: Quick-TRANSfer stage steps exceeded maximum (= 36600)</span></span>
<span><span class="co">#&gt; Warning: Quick-TRANSfer stage steps exceeded maximum (= 36600)</span></span>
<span><span class="co">#&gt; Warning: did not converge in 100000 iterations</span></span>
<span><span class="co">#&gt; Warning: Quick-TRANSfer stage steps exceeded maximum (= 36600)</span></span>
<span><span class="co">#&gt; Warning: Quick-TRANSfer stage steps exceeded maximum (= 36600)</span></span>
<span><span class="co">#&gt; Warning: UNRELIABLE VALUE: One of the 'future.apply' iterations</span></span>
<span><span class="co">#&gt; ('future_lapply-1') unexpectedly generated random numbers without declaring so.</span></span>
<span><span class="co">#&gt; There is a risk that those random numbers are not statistically sound and the</span></span>
<span><span class="co">#&gt; overall results might be invalid. To fix this, specify 'future.seed=TRUE'. This</span></span>
<span><span class="co">#&gt; ensures that proper, parallel-safe random numbers are produced via a parallel</span></span>
<span><span class="co">#&gt; RNG method. To disable this check, use 'future.seed = NULL', or set option</span></span>
<span><span class="co">#&gt; 'future.rng.onMisuse' to "ignore".</span></span></code></pre></div>
<p>This matrix summarizes the results of the analysis of stability of
cluster computation for different values of the maximum number of
groups. Each row corresponds to a tested number of groups and contains
the associated confidence level. The confidence level represents the
proportion of correlation values that are equal to or greater than the
threshold being evaluated. Values closer to 1 indicate a higher
stability of the clustering solution for that number of groups under the
specified confidence level. In other words, values approaching 1 suggest
that a given number of groups yields consistent and reliable cluster
structures.</p>
<p>In this example, the maximum number of groups that yields the most
stable solution is 3. Therefore, we use this value in the
<code>evoregion()</code> computation that follows.</p>
</div>
<div class="section level2">
<h2 id="evoregion">Computing evoregions<a class="anchor" aria-label="anchor" href="#evoregion"></a>
</h2>
<p>Here we use the function <code>calc_evoregion</code>, originally
described in <a href="https://besjournals.onlinelibrary.wiley.com/doi/abs/10.1111/2041-210X.13492" class="external-link">Maestri
and Duarte (2020)</a> and implemented in the {Herodotools} package. This
classification method performs a biogeographical regionalization based
on a <a href="https://doi.org/10.1111/2041-210X.12547" class="external-link">phylogenetic
fuzzy matrix</a>, combined with a <a href="https://bmcgenomdata.biomedcentral.com/articles/10.1186/1471-2156-11-94" class="external-link">Discriminant
Analysis of Principal Components (DAPC) using k-means clustering</a>.
Evoregions represent areas that correspond to centers of lineage
diversification, reflecting historical evolutionary radiations within
clades (Maestri &amp; Duarte, 2020).</p>
<p>To compute evoregions, the user must provide a species occurrence
matrix and a phylogenetic tree. If <code>max.n.clust</code> is not
specified, the <code>evoregion()</code> function automatically selects
the maximum number of clusters using the “elbow” method, as implemented
in the {phyloregion} package (<a href="https://besjournals.onlinelibrary.wiley.com/doi/epdf/10.1111/2041-210X.13478" class="external-link uri">https://besjournals.onlinelibrary.wiley.com/doi/epdf/10.1111/2041-210X.13478</a>).</p>
<p>However, we strongly recommend using the
<code><a href="../reference/find_max_nclust.html">find_max_nclust()</a></code> function to determine the maximum number
of clusters, as demonstrated in the previous section. This approach
identifies the most stable clustering solution and is therefore more
aligned with the methodological rationale of the evoregion
framework.</p>
<p>The <code>seed</code> argument ensures reproducibility: if the
analysis is repeated with the same data and argument settings, the same
numerical labels are assigned to the groups. Without setting a seed, the
group labels (i.e., the cluster numbers) may differ across runs even
when the grouping structure remains the same.</p>
<p>Here, we set <code>max.n.clust = 3</code>, based on the result
obtained in the previous step.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">regions</span> <span class="op">&lt;-</span> </span>
<span>  <span class="fu">Herodotools</span><span class="fu">::</span><span class="fu"><a href="../reference/calc_evoregions.html">calc_evoregions</a></span><span class="op">(</span></span>
<span>  comm <span class="op">=</span> <span class="va">akodon_pa_tree</span>,</span>
<span>  phy <span class="op">=</span> <span class="va">akodon_newick</span>, </span>
<span>  seed <span class="op">=</span> <span class="fl">100</span>, max.n.clust <span class="op">=</span> <span class="fl">3</span></span>
<span>  <span class="op">)</span></span>
<span></span>
<span><span class="va">site_region</span> <span class="op">&lt;-</span> <span class="va">regions</span><span class="op">$</span><span class="va">cluster_evoregions</span> <span class="co"># this is the classification result for each site</span></span></code></pre></div>
<p>We can plot the regions in the map</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">evoregion_df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span></span>
<span>  <span class="va">site_xy</span>, </span>
<span>  <span class="va">site_region</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">r_evoregion</span> <span class="op">&lt;-</span> <span class="fu">terra</span><span class="fu">::</span><span class="fu"><a href="https://rspatial.github.io/terra/reference/rast.html" class="external-link">rast</a></span><span class="op">(</span><span class="va">evoregion_df</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Converting evoregion to a spatial polygon data frame, so it can be plotted</span></span>
<span><span class="va">sf_evoregion</span> <span class="op">&lt;-</span> <span class="fu">terra</span><span class="fu">::</span><span class="fu"><a href="https://rspatial.github.io/terra/reference/as.polygons.html" class="external-link">as.polygons</a></span><span class="op">(</span><span class="va">r_evoregion</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu">sf</span><span class="fu">::</span><span class="fu"><a href="https://r-spatial.github.io/sf/reference/st_as_sf.html" class="external-link">st_as_sf</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="co"># Downloading coastline continents and croping to keep only South America</span></span>
<span><span class="va">coastline</span> <span class="op">&lt;-</span> <span class="fu">rnaturalearth</span><span class="fu">::</span><span class="fu"><a href="https://docs.ropensci.org/rnaturalearth/reference/ne_coastline.html" class="external-link">ne_coastline</a></span><span class="op">(</span>returnclass <span class="op">=</span> <span class="st">"sf"</span><span class="op">)</span></span>
<span><span class="va">map_limits</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>  x <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">95</span>, <span class="op">-</span><span class="fl">30</span><span class="op">)</span>,</span>
<span>  y <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">55</span>, <span class="fl">12</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Assigning the same projection to both spatial objects</span></span>
<span><span class="fu">sf</span><span class="fu">::</span><span class="fu"><a href="https://r-spatial.github.io/sf/reference/st_crs.html" class="external-link">st_crs</a></span><span class="op">(</span><span class="va">sf_evoregion</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu">sf</span><span class="fu">::</span><span class="fu"><a href="https://r-spatial.github.io/sf/reference/st_crs.html" class="external-link">st_crs</a></span><span class="op">(</span><span class="va">coastline</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Colours to plot evoregions</span></span>
<span><span class="va">col_two_hues</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span>  <span class="st">"#3d291a"</span>,</span>
<span>  <span class="st">"#fcc573"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>Finally, plotting evoregions in the map</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span><span class="va">map_evoregion</span> <span class="op">&lt;-</span> </span>
<span>  <span class="va">evoregion_df</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_tile.html" class="external-link">geom_raster</a></span><span class="op">(</span><span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">LONG</span>, y <span class="op">=</span> <span class="va">LAT</span>, fill <span class="op">=</span> <span class="va">site_region</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html" class="external-link">scale_fill_manual</a></span><span class="op">(</span></span>
<span>    name <span class="op">=</span> <span class="st">""</span>, </span>
<span>    labels <span class="op">=</span> <span class="va">LETTERS</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span>,</span>
<span>    values <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/rev.html" class="external-link">rev</a></span><span class="op">(</span><span class="va">col_two_hues</span><span class="op">)</span></span>
<span>  <span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggsf.html" class="external-link">geom_sf</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">coastline</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggsf.html" class="external-link">geom_sf</a></span><span class="op">(</span></span>
<span>    data <span class="op">=</span> <span class="va">sf_evoregion</span>, </span>
<span>    color <span class="op">=</span> <span class="st">"#040400"</span>,</span>
<span>    fill <span class="op">=</span> <span class="cn">NA</span>, </span>
<span>    size <span class="op">=</span> <span class="fl">0.2</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggsf.html" class="external-link">coord_sf</a></span><span class="op">(</span>xlim <span class="op">=</span> <span class="va">map_limits</span><span class="op">$</span><span class="va">x</span>, ylim <span class="op">=</span> <span class="va">map_limits</span><span class="op">$</span><span class="va">y</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">""</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_bw</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">xlab</a></span><span class="op">(</span><span class="st">"Longitude"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ylab</a></span><span class="op">(</span><span class="st">"Latitude"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span></span>
<span>        legend.position <span class="op">=</span> <span class="st">"bottom"</span>,</span>
<span>        plot.margin <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/grid/unit.html" class="external-link">unit</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.1</span>, <span class="fl">0.1</span>, <span class="fl">0.1</span>, <span class="fl">0.1</span><span class="op">)</span>, <span class="st">"mm"</span><span class="op">)</span>,</span>
<span>        legend.text <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">12</span><span class="op">)</span>, </span>
<span>        axis.text <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">7</span><span class="op">)</span>,</span>
<span>        axis.title.x <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">11</span><span class="op">)</span>,</span>
<span>        axis.title.y <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">11</span><span class="op">)</span></span>
<span>        <span class="op">)</span></span>
<span></span>
<span><span class="va">map_evoregion</span></span></code></pre></div>
<p><img src="building-evoregions_files/figure-html/unnamed-chunk-7-1.png" width="700">
The output from <code>evoregion()</code> indicates the presence of two
distinct regions. However, not all cells share the same degree of
affiliation with the region to which they were assigned. Cells with high
affiliation values represent assemblages that are highly similar to
other cells within the same region. Conversely, cells with low
affiliation values correspond to areas with <a href="https://besjournals.onlinelibrary.wiley.com/doi/full/10.1111/2041-210X.13492" class="external-link">high
turnover</a>, that is, areas where multiple colonization events by
different lineages have occurred (Maestri &amp; Duarte, 2020).</p>
<p>We can calculate the affiliation of each cell to its corresponding
region with the function <code>calc_affiliation_evoreg</code></p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co"># distance matrix using 4 significant PCPS axis accordingly to the 5% threshold </span></span>
<span><span class="va">dist_phylo_PCPS</span> <span class="op">&lt;-</span> <span class="fu">vegan</span><span class="fu">::</span><span class="fu"><a href="https://vegandevs.github.io/vegan/reference/vegdist.html" class="external-link">vegdist</a></span><span class="op">(</span><span class="va">vec_bray</span><span class="op">[</span>, <span class="fl">1</span><span class="op">:</span><span class="va">thresh_bray</span><span class="op">]</span>, method <span class="op">=</span> <span class="st">"euclidean"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># calculating affiliation values for each assemblage </span></span>
<span><span class="va">afi</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/calc_affiliation_evoreg.html">calc_affiliation_evoreg</a></span><span class="op">(</span>phylo.comp.dist <span class="op">=</span> <span class="va">dist_phylo_PCPS</span>,</span>
<span>                          groups <span class="op">=</span> <span class="va">site_region</span><span class="op">)</span> </span>
<span></span>
<span><span class="co"># binding the information in a data frame</span></span>
<span><span class="va">sites</span> <span class="op">&lt;-</span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/bind_cols.html" class="external-link">bind_cols</a></span><span class="op">(</span><span class="va">site_xy</span>, site_region <span class="op">=</span>  <span class="va">site_region</span>, <span class="va">afi</span><span class="op">)</span></span></code></pre></div>
<p>Now we can map the evoregions along with the degree of affiliation
for each cell. The affiliation values are represented by the intensity
of the colors: cells with high affiliation appear in strong, saturated
colors, while cells with low affiliation are shown in more faded colors.
In other words, the more faded a cell’s color, the weaker its
affiliation to the assigned evoregion.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">map_joint_evoregion_afilliation</span> <span class="op">&lt;-</span> </span>
<span>   <span class="va">evoregion_df</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> </span>
<span>   <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> </span>
<span>   <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_tile.html" class="external-link">geom_raster</a></span><span class="op">(</span><span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">LONG</span>, y <span class="op">=</span> <span class="va">LAT</span>, fill <span class="op">=</span> <span class="va">site_region</span><span class="op">)</span>, </span>
<span>               alpha <span class="op">=</span> <span class="va">sites</span><span class="op">[</span>, <span class="st">"afilliation"</span><span class="op">]</span><span class="op">)</span> <span class="op">+</span> </span>
<span>   <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html" class="external-link">scale_fill_manual</a></span><span class="op">(</span></span>
<span>     name <span class="op">=</span> <span class="st">""</span>, </span>
<span>     labels <span class="op">=</span> <span class="va">LETTERS</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">2</span><span class="op">]</span>,</span>
<span>     values <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/rev.html" class="external-link">rev</a></span><span class="op">(</span><span class="va">col_two_hues</span><span class="op">)</span></span>
<span>   <span class="op">)</span> <span class="op">+</span></span>
<span>   <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggsf.html" class="external-link">geom_sf</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">coastline</span>, size <span class="op">=</span> <span class="fl">0.4</span><span class="op">)</span> <span class="op">+</span></span>
<span>   <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggsf.html" class="external-link">geom_sf</a></span><span class="op">(</span></span>
<span>     data <span class="op">=</span> <span class="va">sf_evoregion</span>, </span>
<span>     color <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/rev.html" class="external-link">rev</a></span><span class="op">(</span><span class="va">col_two_hues</span><span class="op">)</span>,</span>
<span>     fill <span class="op">=</span> <span class="cn">NA</span>, </span>
<span>     size <span class="op">=</span> <span class="fl">0.7</span><span class="op">)</span> <span class="op">+</span></span>
<span>   <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggsf.html" class="external-link">coord_sf</a></span><span class="op">(</span>xlim <span class="op">=</span> <span class="va">map_limits</span><span class="op">$</span><span class="va">x</span>, ylim <span class="op">=</span> <span class="va">map_limits</span><span class="op">$</span><span class="va">y</span><span class="op">)</span> <span class="op">+</span></span>
<span>   <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">""</span><span class="op">)</span> <span class="op">+</span> </span>
<span>   <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/guides.html" class="external-link">guides</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/guide_legend.html" class="external-link">guide_legend</a></span><span class="op">(</span>direction <span class="op">=</span> <span class="st">"vertical"</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>   <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_bw</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">xlab</a></span><span class="op">(</span><span class="st">"Longitude"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ylab</a></span><span class="op">(</span><span class="st">"Latitude"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span></span>
<span>        legend.position <span class="op">=</span> <span class="st">"bottom"</span>,</span>
<span>        plot.margin <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/grid/unit.html" class="external-link">unit</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.1</span>, <span class="fl">0.1</span>, <span class="fl">0.1</span>, <span class="fl">0.1</span><span class="op">)</span>, <span class="st">"mm"</span><span class="op">)</span>,</span>
<span>        legend.text <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">10</span><span class="op">)</span>, </span>
<span>        axis.text <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">8</span><span class="op">)</span>,</span>
<span>        axis.title.x <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">10</span><span class="op">)</span>,</span>
<span>        axis.title.y <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">10</span><span class="op">)</span></span>
<span>        <span class="op">)</span></span>
<span><span class="co">#&gt; Warning: Guides provided to `guides()` must be named.</span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> All guides are unnamed.</span></span>
<span></span>
<span><span class="va">map_joint_evoregion_afilliation</span></span></code></pre></div>
<p><img src="building-evoregions_files/figure-html/mapAffiliation-1.png" width="700"></p>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Gabriel Nakamura, Arthur Rodrigues, André Luza, Vanderlei Debastiani, Renan Maestri, Leandro Duarte.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer>
</div>





  </body>
</html>
